tests/selenium/conftest.py

```diff
@@ -0,0 +1,341 @@
+import pytest
+
+from dash.testing.browser import Browser
+from preconditions import preconditions
+from selenium.webdriver.common.action_chains import ActionChains
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.wait import WebDriverWait
+
+_validate_col = lambda col: (isinstance(col, str) and len(col) > 0) or (isinstance(col, int) and col >= 0)
+_validate_col_id = lambda col_id: isinstance(col_id, str) and len(col_id) > 0
+_validate_id = lambda id: isinstance(id, str) and len(id) > 0
+_validate_key = lambda key: isinstance(key, str) and len(key) == 1
+_validate_keys = lambda keys: isinstance(keys, str) and len(keys) > 0
+_validate_mixin = lambda mixin: isinstance(mixin, DataTableMixin)
+_validate_row = lambda row: isinstance(row, int) and row >= 0
+_validate_state = lambda state: state in [_READY, _LOADING, _ANY]
+_validate_target = lambda target: isinstance(target, DataTableFacade)
+
+_READY = '.dash-spreadsheet:not(.dash-loading)'
+_LOADING = '.dash-spreadsheet.dash-loading'
+_ANY = '.dash-spreadsheet'
+_TIMEOUT = 10
+
+class DataTableContext:
+    @preconditions(_validate_target)
+    def __init__ (self, target):
+        self.target = target
+
+    def __enter__(self):
+        return self.target
+
+    def __exit__(self, type, value, traceback):
+        return
+
+
+class HoldKeyContext:
+    @preconditions(_validate_mixin, _validate_key)
+    def __init__ (self, mixin, key):
+        self.mixin = mixin
+        self.key = key
+
+    def __enter__(self):
+        ActionChains(self.mixin.driver).key_down(self.key).perform()
+
+    def __exit__(self, type, value, traceback):
+        ActionChains(self.mixin.driver).key_up(self.key).perform()
+
+
+class DataTableCellFacade(object):
+    @preconditions(_validate_id, _validate_mixin)
+    def __init__(self, id, mixin):
+        self.id = id
+        self.mixin = mixin
+
+    @preconditions(_validate_row, _validate_col, _validate_state)
+    def _get_cell_value(self, row, col, selector, state = _ANY):
```

Comment:
If we ever add linting to our tests it'll tell you default args get no spaces `state=_ANY`.

Now might not be a bad time to do that actually, since there isn't that much existing Python test code that would need cleaning up.

Reply:
Agreed

Reply:
https://github.com/plotly/dash-table/pull/696/commits/3d9cf3701730dd216cf0af1c41a4ff67ffba4153